pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/SchemaRegistryServiceImpl.java

```diff
@@ -181,56 +183,85 @@ public CompletableFuture<List<CompletableFuture<SchemaAndMetadata>>> getAllSchem
-                        .setUser(schema.getUser())
-                        .setDeleted(false)
-                        .setTimestamp(clock.millis())
-                        .addAllProps(toPairs(schema.getProps()))
-                        .build();
-
-                long start = this.clock.millis();
-
-                return schemaStorage
-                        .put(schemaId, info.toByteArray(), context)
-                        .whenComplete((__, t) -> {
-                            if (t != null) {
-                                log.error("[{}] Put schema failed", schemaId);
-                                this.stats.recordPutFailed(schemaId);
-                            } else {
-                                if (log.isDebugEnabled()) {
-                                    log.debug("[{}] Put schema finished", schemaId);
-                                }
-                                this.stats.recordPutLatency(schemaId, this.clock.millis() - start);
-                            }
-                        });
-
-            });
-
-        }));
+        CompletableFuture<SchemaVersion> persistentFuture = new CompletableFuture<>();
+        return trimDeletedSchemaAndGetList(schemaId)
+                .thenCompose(schemaAndMetadataList ->
+                        getSchemaVersionBySchemaData(schemaAndMetadataList, schema)
+                                .thenCompose(schemaVersion -> {
+                                    if (schemaVersion != null) {
+                                        if (log.isDebugEnabled()) {
+                                            log.debug("[{}] Schema is already exists", schemaId);
+                                        }
+                                        return CompletableFuture.completedFuture(schemaVersion);
+                                    }
+                                    CompletableFuture<Void> checkCompatibilityFuture = new CompletableFuture<>();
+                                    if (schemaAndMetadataList.size() != 0) {
+                                        if (isTransitiveStrategy(strategy)) {
+                                            checkCompatibilityFuture = checkCompatibilityWithAll(schemaId, schema,
+                                                    strategy, schemaAndMetadataList);
+                                        } else {
+                                            checkCompatibilityFuture = checkCompatibilityWithLatest(schemaId,
+                                                    schema, strategy);
+                                        }
+                                    } else {
+                                        checkCompatibilityFuture.complete(null);
+                                    }
+                                    return checkCompatibilityFuture.thenCompose(v -> {
+                                        byte[] context = hashFunction.hashBytes(schema.getData()).asBytes();
```

Comment:
it's better to move line-186 to this block.

Reply:
fixed